### ISSUE TITLE
I1620: Typewriter screen wrap is messed up

### ISSUE THREAD

**Issue by DaveWise2** [simh/simh#438]
- #### Context

When running under Windows 7, typewriter output wraps at column 80 like it should, but then it wraps again after displaying one character.  There are two problems.  The commit 316565c45b94872915d4e1d1ddaaa32590bb27b3, "CONSOLE: Really support Windows 'ANSI' escape sequences in console sessions", added to sim_os_ttrun() in sim_console.c a SetConsoleMode() argument of ENABLE_VIRTUAL_TERMINAL_PROCESSING, which only exists in recent Windows 10 builds.  When I1620.exe is run under earlier builds and versions, such as Windows 7, the SetConsoleMode() fails, which leaves the mode at its default value of 3, i.e. ENABLE_WRAP_AT_EOL_OUTPUT | ENABLE_PROCESSED_OUTPUT.  So Windows wraps, and SIMH wraps too.  You have to omit ENABLE_VIRTUAL_TERMINAL_PROCESSING from the argument when it's not supported, so that you can turn off ENABLE_WRAP_AT_EOL_OUTPUT.

Also, tto_write() in i1620_tty.c outputs 81 characters before it wraps, because tto_col starts at 0 but the endpoint compare is `if (tto_col > TTO_COLMAX)` and TTO_COLMAX is 80.  I fixed it locally by changing > to >= , but maybe this misses other sore spots.

- #### the output of "sim> SHOW VERSION" while running the simulator which is having the issue

sim> show ver
IBM 1620 simulator V4.0-0 Beta
    Simulator Framework Capabilities:
        32b data
        32b addresses
        no Ethernet
        Idle/Throttling support is available
        Virtual Hard Disk (VHD) support
        RAW disk and CD/DVD ROM support
        FrontPanel API Version 4
    Host Platform:
        Compiler: Microsoft Visual C++ 15.00.30729.01
        Simulator Compiled as C arch: x86 (Release Build) on Apr 24 2017 at 13:48:57
        Memory Access: Little Endian
        Memory Pointer Size: 32 bits
        Large File (>2GB) support
        SDL Video support: No Video Support
        PCRE RegEx support for EXPECT commands
        OS clock resolution: 1ms
        Time taken by msleep(1): 1ms
        OS: Microsoft Windows [Version 6.1.7601]
        Architecture: x86 on AMD64, Processors: 36
        Processor Id: Intel64 Family 6 Model 63 Stepping 2, GenuineIntel, Level: 6, Revision: 3f02
        git commit id: b5a5aec5
sim>

- #### how you built the simulator or that you're using prebuilt binaries

Current prebuilt binary.

- #### the simulator configuration file (or commands) which were used when the problem occurred.

None

- #### the expected behavior and the actual behavior

Expected:

sim> d 0 dnty 19000
sim> d 12 h
sim> go 0
0000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000
HALT instruction, PC: 24 (00)
sim>

Actual:

sim> d 0 dnty 19000
sim> d 12 h
sim> go 0
0000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
00000000000000000000000000000000000000000000000000000000000000000000000000000000
0
000000000000000000000000000000000000000000000000000
HALT instruction, PC: 24 (00)
sim>

The truncation of the first line is another problem, which I will report separately.

- #### you may also need to provide specific pointers to data files that may be necessary to demonstrate the problem

**Comment by markpizz** [comment]
My inclination is to start with this detail:
>Also, tto_write() in i1620_tty.c outputs 81 characters before it wraps, because tto_col starts at 0 but the endpoint compare is if (tto_col > TTO_COLMAX) and TTO_COLMAX is 80. I fixed it locally by changing > to >= , but maybe this misses other sore spots.

There may indeed be other sore spots, but what is the output produced when:
```
if (tto_col > TTO_COLMAX)
```
is changed to:
```
if (tto_col => TTO_COLMAX)
```

Do you see reasonable output:
```
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
```
OR do you see:
```
00000000000000000000000000000000000000000000000000000000000000000000000000000000

00000000000000000000000000000000000000000000000000000000000000000000000000000000

00000000000000000000000000000000000000000000000000000000000000000000000000000000
```

**Comment by DaveWise2** [comment]
Changing only the tto_col comparison to TTO_COLMAX, I get the second, different from stock but still incorrect display in your post above.  That's because Windows and SIMH are both wrapping.  To get Windows to stop wrapping, you have to successfully execute SetConsoleMode() with ENABLE_WRAP_AT_EOL_OUTPUT turned off.  Under Windows 7, the existing call `SetConsoleMode(std_output, ENABLE_VIRTUAL_TERMINAL_PROCESSING|ENABLE_PROCESSED_OUTPUT)` fails, because Windows 7 doesn't support ENABLE_VIRTUAL_TERMINAL_PROCESSING, leaving the console mode at its default setting, which on my system is 3, ENABLE_WRAP_AT_EOL_OUTPUT | ENABLE_PROCESSED_OUTPUT.

If I change the above code to
```
        if (!SetConsoleMode(std_output, ENABLE_VIRTUAL_TERMINAL_PROCESSING|ENABLE_PROCESSED_OUTPUT))
            SetConsoleMode(std_output, ENABLE_PROCESSED_OUTPUT);
```
I get your first, reasonable output.

**Comment by markpizz** [comment]
OK.  That is a generic bug in the console layer which was introduced where you noted.  It is now fixed.

As for the 81 character question, What was the real output device for one of these systems? Just guessing, but could the starting character be some line printer format code?

**Comment by DaveWise2** [comment]
Thanks, Mark.

No, there are no hidden format codes or other non-printing characters.  The 1620 Model I Console is a Model B typewriter with contacts and solenoids under the keys.  (The Model II Console is a Selectric with contacts and solenoids.)  Except for the character set, and being electrically controlled, it is just another typewriter.  Every cycle of WATY/WNTY/DNTY data transfer moves the carriage one tick forward.  The bell rings 10 ticks before right margin.  When the carriage strikes the right margin, the next cycle causes a carriage-return (with linefeed) before kicking its typebar.

The margins are adjustable up to the length of the carriage.  The standard carriage is 80 characters.  The typical margin setting is 1 and 80 (i.e. 0 and 79.)  A few systems - like mine - were sold with the typewriter on a separate stand, customized with a long carriage.  (It has to be on a stand because if it were mounted in the desk as usual, the carriage would strike the CPU cabinet.  It also has a guard on the right to prevent it from striking the operator in a tender spot!)  But SIMH's text-mode console screen is limited to 80 characters anyway, at least under Windows, at least as currently coded, so if right margin is not adjustable, it should permit exactly 80 characters per line.

**Comment by DaveWise2** [comment]
Mark, is TTO_COLMAX (a) the number of columns, or (b) the maximum column number at which you can print?  And if (b), is it zero-based or one-based?

**Comment by DaveWise2** [comment]
Mark, is TTO_COLMAX the number of columns?  Or is it the maximum permitted column?  Zero-based or one-based?

**Comment by markpizz** [comment]
> Mark, is TTO_COLMAX the number of columns? Or is it the maximum permitted column? Zero-based or one-based?

Well, those seem to be the key questions here.

This is all the more complicated by the support for tab stops.

It would seem natural for all of the indexing activities to be zero-based, but I didn't write this code and never looked closely at it prior to this discussion.

**Comment by DaveWise2** [comment]
I'll try to ascertain how each use is treating it.  Maybe one case is a majority.  What do you think about the 80-character limit I mentioned above?  I don't want to be accused of eighty-column mind, but this _is_  IBM after all.

**Comment by markpizz** [comment]
There really isn't an 80 column limit imposed by a simh console.  In your case, simh is running in a Windows command prompt window which defaults to 80 columns wide, but you can change that to any thing you like.  This doesn't really matter since that is the mixed use mode (i.e. console mixed with simh control).  

The interesting case is the telnet session.  PuTTY defaults to emulate an 80 column emulator, but that can also be changed to any reasonable dimension you want.  On your real hardware with the 'long carriage', what is the maximum line length of that carriage?

**Comment by DaveWise2** [comment]
I thought that might be the case, but it seems to me it would be messy getting all the configurations lined up.  I'll figure out what my carriage supports and post it tomorrow.  132 rings a bell.  (For what it's worth, the entire time I've owned the machine - since 1978 - it's had an 80-column platen installed in the middle of that long carriage.)

**Comment by markpizz** [comment]
Just curious, what mechanism actually moves the paper?

**Comment by markpizz** [comment]
> The margins are adjustable up to the length of the carriage. The standard carriage is 80 characters. The typical margin setting is 1 and 80 (i.e. 0 and 79.) A few systems - like mine - were sold with the typewriter on a separate stand, customized with a long carriage. (It has to be on a stand because if it were mounted in the desk as usual, the carriage would strike the CPU cabinet. It also has a guard on the right to prevent it from striking the operator in a tender spot!) 

It sounds like you're describing a moving carriage (like very old type writers) as opposed to a moving type head that the Selectric Typewriters (and Terminals) had.  Are you?

**Comment by dlrourke** [comment]
Replying for Mr. Wise:

 

The original Selectric typewriter was introduced in 1961, but the IBM 1620 in 1959.  The “console” on the 1620 was a moving carriage (platen may be a better word) typewriter.  I believe in an earlier email Mr. Wise said that on the later 1620’s a Selectric-like console was used.

**Comment by markpizz** [comment]
Given a moving carriage device, what type of paper was used?  Did you have individual sheets that had to be manually inserted or did it feed continuously from either a roll or a box?

**Comment by DaveWise2** [comment]
<Are you?> Yes.

The 1620 Model I (like I have) uses an IBM Model B, which has a basket of typebars and a moving carriage with removable platen.  It can handle individual hand-fed sheets of paper, or if the platen is a pinfeed model (which I have), it can also use continuous pinfeed paper from a box like the old "green bar" stuff all of us of a certain age remember.  That is what I use day to day.  I don't have direct personal experience with the Model II, but it uses a Selectric, which has a stationary platen and "infuriated golfball".  I have no doubt it could take a pinfeed platen too.

**Comment by markpizz** [comment]
So, the with the pinfeed paper coming up from a box below, isn't there considerable drag on that paper as the carriage lurches back on a CR-LF?

**Comment by DaveWise2** [comment]
You have to center the box, but there's enough distance between it and the typewriter for the paper to flex without tearing.

**Comment by markpizz** [comment]
Of course that would be more exciting with a wide (132) character platen...

**Comment by DaveWise2** [comment]
Yep  =8-0

**Comment by DaveWise2** [comment]
My carriage is 180 chars wide!  I'm not sure one can utilize all of that with a pinfeed platen, but at least 160 for sure.

**Comment by DaveWise2** [comment]
Original complaint has been addressed in issue #439.  Maybe someday we can make the virtual carriage wide, with adjustable margins, but I'll open a separate feature request then.
